File: forge-core/internal/kerneldecisioncontract/atom.go
package kerneldecisioncontract

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"sort"
)

var sourceTypes = map[string]map[string]bool{
	"artifact":              allAtomTypes(),
	"artifact_receipt":      atomTypeSet("evidence", "object", "observation"),
	"capability_invocation": atomTypeSet("actor", "operation"),
	"cognitive_atom_v1":     atomTypeSet("assumption", "constraint", "decision", "fact", "hypothesis", "inference", "unknown"),
	"evidence_record":       atomTypeSet("evidence", "observation"),
	"execution_receipt":     atomTypeSet("actor", "evidence", "observation"),
	"interaction_event":     atomTypeSet("actor", "evidence", "object", "observation", "operation"),
	"work_intent":           atomTypeSet("acceptance", "constraint", "goal", "preference", "risk", "unknown"),
}

var hardnessByType = map[string]map[string]bool{
	"acceptance": atomTypeSet("advisory", "preferred", "required"),
	"actor":      atomTypeSet("none"), "assumption": atomTypeSet("advisory", "none"),
	"constraint": atomTypeSet("advisory", "contract", "invariant", "preferred", "required"),
	"decision":   atomTypeSet("advisory", "required"), "evidence": atomTypeSet("none"),
	"fact": atomTypeSet("none"), "goal": atomTypeSet("advisory", "preferred", "required"),
	"hypothesis": atomTypeSet("advisory", "none"), "inference": atomTypeSet("advisory", "none"),
	"object": atomTypeSet("none"), "observation": atomTypeSet("none"),
	"operation": atomTypeSet("none"), "preference": atomTypeSet("advisory", "preferred"),
	"risk": atomTypeSet("advisory", "none"), "unknown": atomTypeSet("advisory", "none"),
}

func atomTypeSet(values ...string) map[string]bool {
	result := make(map[string]bool, len(values))
	for _, value := range values {
		result[value] = true
	}
	return result
}

func allAtomTypes() map[string]bool {
	return atomTypeSet("acceptance", "actor", "assumption", "constraint", "decision",
		"evidence", "fact", "goal", "hypothesis", "inference", "object", "observation",
		"operation", "preference", "risk", "unknown")
}

func validateProposition(value Proposition) error {
	if err := identifier(value.Predicate, "proposition.predicate"); err != nil {
		return err
	}
	if err := identifier(value.Subject, "proposition.subject"); err != nil {
		return err
	}
	switch value.ObjectType {
	case "artifact_ref":
		member, ok := value.ObjectValue.(string)
		if !ok {
			return fmt.Errorf("artifact_ref proposition object must be text")
		}
		return identifier(member, "proposition.object_value")
	case "string":
		member, ok := value.ObjectValue.(string)
		if !ok {
			return fmt.Errorf("string proposition object must be text")
		}
		return text(member, "proposition.object_value", maxStringBytes)
	case "boolean":
		if _, ok := value.ObjectValue.(bool); !ok {
			return fmt.Errorf("boolean proposition object has wrong type")
		}
	case "integer":
		if !signedInteger(value.ObjectValue) {
			return fmt.Errorf("integer proposition object has wrong type")
		}
	case "null":
		if value.ObjectValue != nil {
			return fmt.Errorf("null proposition object must be null")
		}
	default:
		return fmt.Errorf("proposition.object_type is unsupported")
	}
	return nil
}

func signedInteger(value any) bool {
	switch number := value.(type) {
	case json.Number:
		_, err := number.Int64()
		return err == nil
	case int, int8, int16, int32, int64:
		return true
	case uint:
		return uint64(number) <= math.MaxInt64
	case uint8, uint16, uint32:
		return true
	case uint64:
		return number <= math.MaxInt64
	default:
		return false
	}
}

func validateAtomSemantics(value *CognitiveAtom) error {
	if value.AtomAPIInvalid() {
		return fmt.Errorf("CognitiveAtom constants differ")
	}
	if !allAtomTypes()[value.AtomType] {
		return fmt.Errorf("atom_type is unsupported")
	}
	if err := validateAtomMembers(value); err != nil {
		return err
	}
	return validateAtomState(value)
}

func (value *CognitiveAtom) AtomAPIInvalid() bool {
	return value.APIVersion != atomAPI || value.Canonicalization != canonicalization ||
		value.EffectiveHardness != "none" || value.InstructionAllowed || value.Kind != atomKind
}

func validateAtomMembers(value *CognitiveAtom) error {
	if err := validateAttestations(value.Attestations); err != nil {
		return err
	}
	if err := validateBindings(value.Bindings); err != nil {
		return err
	}
	if err := validateTask(value.TaskBinding); err != nil {
		return err
	}
	if err := validateProposition(value.Proposition); err != nil {
		return err
	}
	if value.Scope.Project != value.TaskBinding.ProjectID {
		return fmt.Errorf("scope.project must equal task_binding.project_id")
	}
	if err := text(value.Scope.Project, "scope.project", maxShortBytes); err != nil {
		return err
	}
	if err := validateScopeOptional(value); err != nil {
		return err
	}
	if err := validateAuthority(value.DeclaredAuthority); err != nil {
		return err
	}
	return validateSource(value.Source, value.AtomType)
}

func validateScopeOptional(value *CognitiveAtom) error {
	if value.Scope.Module != nil {
		if err := text(*value.Scope.Module, "scope.module", maxShortBytes); err != nil {
			return err
		}
	}
	if value.Scope.Object != nil {
		if err := text(*value.Scope.Object, "scope.object", maxShortBytes); err != nil {
			return err
		}
		if *value.Scope.Object != value.Proposition.Subject {
			return fmt.Errorf("scope.object must equal proposition.subject")
		}
	}
	return nil
}

func validateAtomState(value *CognitiveAtom) error {
	if value.Validity.ValidFromUnixMS < 0 || value.Validity.ValidUntilUnixMS != nil &&
		(*value.Validity.ValidUntilUnixMS < 0 || *value.Validity.ValidUntilUnixMS <= value.Validity.ValidFromUnixMS) {
		return fmt.Errorf("CognitiveAtom validity is invalid")
	}
	requiresConfidence := oneOf(value.AtomType, "assumption", "hypothesis", "inference")
	validConfidence := value.ConfidenceMicros != nil && *value.ConfidenceMicros >= 0 && *value.ConfidenceMicros <= 1000000
	if requiresConfidence != validConfidence || !requiresConfidence && value.ConfidenceMicros != nil {
		return fmt.Errorf("confidence_micros presence does not match atom_type")
	}
	if err := validateHardness(value); err != nil {
		return err
	}
	return validateEpistemic(value)
}

func validateHardness(value *CognitiveAtom) error {
	legacy := value.Source.SourceKind == "cognitive_atom_v1"
	if legacy {
		if value.DeclaredHardness != "none" || value.DeclaredAuthority.AuthorityKind != "none" {
			return fmt.Errorf("legacy source requires none hardness and authority")
		}
		return nil
	}
	if !hardnessByType[value.AtomType][value.DeclaredHardness] {
		return fmt.Errorf("declared_hardness is not admitted by atom_type")
	}
	authority := value.DeclaredAuthority.AuthorityKind
	if value.DeclaredHardness == "none" && authority != "none" {
		return fmt.Errorf("none hardness requires none authority")
	}
	if oneOf(value.DeclaredHardness, "contract", "invariant") && authority != "contract_artifact" {
		return fmt.Errorf("contract/invariant hardness requires contract artifact")
	}
	if value.DeclaredHardness == "required" && authority == "none" {
		return fmt.Errorf("required hardness requires declared authority")
	}
	if value.DeclaredHardness == "required" && value.AtomType == "decision" &&
		!oneOf(authority, "approval_record", "architecture_decision") {
		return fmt.Errorf("required decision requires ADR or Approval")
	}
	return nil
}

func validateEpistemic(value *CognitiveAtom) error {
	if value.Source.SourceKind != "cognitive_atom_v1" {
		if value.EpistemicState != "declared" {
			return fmt.Errorf("nonlegacy epistemic_state must be declared")
		}
		return nil
	}
	allowed := map[string][]string{"assumption": {"open", "testing"}, "constraint": {"candidate"},
		"decision": {"proposed"}, "fact": {"candidate", "contested"},
		"hypothesis": {"open", "testing"}, "inference": {"candidate"},
		"unknown": {"investigating", "open"}}
	if !oneOf(value.EpistemicState, allowed[value.AtomType]...) {
		return fmt.Errorf("legacy epistemic_state is outside ADR-0047 states")
	}
	return nil
}

func atomDigest(value *CognitiveAtom) (string, error) {
	blank := *value
	blank.AtomID, blank.AtomSHA256 = "", ""
	if err := validateAtomBody(&blank, true); err != nil {
		return "", err
	}
	raw, err := canonicalBytes(&blank, maxAtomBytes)
	if err != nil {
		return "", err
	}
	digest := sha256.New()
	digest.Write(atomDomain)
	digest.Write(raw)
	return hex.EncodeToString(digest.Sum(nil)), nil
}

func validateAtomBody(value *CognitiveAtom, blank bool) error {
	if value == nil {
		return fmt.Errorf("CognitiveAtom is nil")
	}
	if err := identity(value.AtomID, value.AtomSHA256, atomPrefix, "atom", blank); err != nil {
		return err
	}
	return validateAtomSemantics(value)
}

func ValidateCognitiveAtom(value *CognitiveAtom) error {
	if err := validateAtomBody(value, false); err != nil {
		return err
	}
	digest, err := atomDigest(value)
	if err != nil {
		return err
	}
	if value.AtomSHA256 != digest {
		return fmt.Errorf("atom_sha256 does not match canonical preimage")
	}
	_, err = canonicalBytes(value, maxAtomBytes)
	return err
}

func SealCognitiveAtom(value *CognitiveAtom) (*CognitiveAtom, error) {
	if value == nil || value.AtomID != "" || value.AtomSHA256 != "" {
		return nil, fmt.Errorf("sealing CognitiveAtom requires blank identity")
	}
	sealed, err := cloneValue(value)
	if err != nil {
		return nil, err
	}
	digest, err := atomDigest(sealed)
	if err != nil {
		return nil, err
	}
	sealed.AtomID, sealed.AtomSHA256 = atomPrefix+digest, digest
	return sealed, ValidateCognitiveAtom(sealed)
}

func DecodeCognitiveAtom(raw []byte) (*CognitiveAtom, error) {
	var value CognitiveAtom
	if err := decodeTyped(raw, maxAtomBytes, &value); err != nil {
		return nil, err
	}
	return &value, ValidateCognitiveAtom(&value)
}

func sortAtoms(values []CognitiveAtom) error {
	if values == nil || len(values) == 0 || len(values) > maxAtoms {
		return fmt.Errorf("cognitive_atoms cardinality must be 1..%d", maxAtoms)
	}
	ids := make([]string, len(values))
	for index := range values {
		if err := ValidateCognitiveAtom(&values[index]); err != nil {
			return fmt.Errorf("cognitive_atoms[%d]: %w", index, err)
		}
		ids[index] = values[index].AtomID
	}
	if !sort.StringsAreSorted(ids) || duplicateStrings(ids) {
		return fmt.Errorf("cognitive_atoms must be strictly atom-id sorted and unique")
	}
	return nil
}

File: forge-runtime/crates/domain/src/kernel_decision_contract/atom.rs
use serde_json::Value;
use sha2::{Digest, Sha256};

use super::{
    ATOM_API, ATOM_DOMAIN, ATOM_KIND, ATOM_PREFIX, CANONICALIZATION, CognitiveAtom,
    KernelDecisionContractError, MAX_ATOM_BYTES, invalid,
    primitives::{attestations, bindings, identifier, identity, task, text},
    source::{atom_type_member, authority, source},
    wire,
};

fn proposition(value: &super::Proposition) -> Result<(), KernelDecisionContractError> {
    identifier(&value.predicate, "proposition.predicate")?;
    identifier(&value.subject, "proposition.subject")?;
    match (value.object_type.as_str(), &value.object_value) {
        ("artifact_ref", Value::String(text)) => identifier(text, "proposition.object_value"),
        ("string", Value::String(member)) => text(member, "proposition.object_value", 16_384),
        ("boolean", Value::Bool(_)) | ("null", Value::Null) => Ok(()),
        ("integer", Value::Number(number)) if number.as_i64().is_some() => Ok(()),
        ("artifact_ref" | "string" | "boolean" | "integer" | "null", _) => Err(invalid(
            "proposition.object_value does not match object_type",
        )),
        _ => Err(invalid("proposition.object_type is unsupported")),
    }
}

fn scope(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    text(&value.scope.project, "scope.project", 160)?;
    if value.scope.project != value.task_binding.project_id {
        return Err(invalid("scope.project must equal task_binding.project_id"));
    }
    if let Some(module) = &value.scope.module {
        text(module, "scope.module", 160)?;
    }
    if let Some(object) = &value.scope.object {
        text(object, "scope.object", 160)?;
        if object != &value.proposition.subject {
            return Err(invalid("scope.object must equal proposition.subject"));
        }
    }
    Ok(())
}

fn confidence(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    let required = matches!(
        value.atom_type.as_str(),
        "assumption" | "hypothesis" | "inference"
    );
    let valid = value
        .confidence_micros
        .is_some_and(|number| (0..=1_000_000).contains(&number));
    if required == valid && (required || value.confidence_micros.is_none()) {
        Ok(())
    } else {
        Err(invalid(
            "confidence_micros presence does not match atom_type",
        ))
    }
}

fn admitted_hardness(atom_type: &str, hardness: &str) -> bool {
    match atom_type {
        "acceptance" | "goal" => matches!(hardness, "advisory" | "preferred" | "required"),
        "actor" | "evidence" | "fact" | "object" | "observation" | "operation" => {
            hardness == "none"
        }
        "assumption" | "hypothesis" | "inference" | "risk" | "unknown" => {
            matches!(hardness, "advisory" | "none")
        }
        "constraint" => matches!(
            hardness,
            "advisory" | "contract" | "invariant" | "preferred" | "required"
        ),
        "decision" => matches!(hardness, "advisory" | "required"),
        "preference" => matches!(hardness, "advisory" | "preferred"),
        _ => false,
    }
}

fn hardness(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    let authority_kind = value.declared_authority.authority_kind.as_str();
    if value.source.source_kind == "cognitive_atom_v1" {
        if value.declared_hardness == "none" && authority_kind == "none" {
            return Ok(());
        }
        return Err(invalid(
            "legacy source requires none hardness and authority",
        ));
    }
    if !admitted_hardness(&value.atom_type, &value.declared_hardness) {
        return Err(invalid("declared_hardness is not admitted by atom_type"));
    }
    match value.declared_hardness.as_str() {
        "none" if authority_kind != "none" => Err(invalid("none hardness requires none authority")),
        "contract" | "invariant" if authority_kind != "contract_artifact" => Err(invalid(
            "contract/invariant hardness requires contract artifact",
        )),
        "required" if authority_kind == "none" => {
            Err(invalid("required hardness requires declared authority"))
        }
        "required"
            if value.atom_type == "decision"
                && !matches!(authority_kind, "approval_record" | "architecture_decision") =>
        {
            Err(invalid("required decision requires ADR or Approval"))
        }
        _ => Ok(()),
    }
}

fn epistemic(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    if value.source.source_kind != "cognitive_atom_v1" {
        if value.epistemic_state == "declared" {
            return Ok(());
        }
        return Err(invalid("nonlegacy epistemic_state must be declared"));
    }
    let valid = match value.atom_type.as_str() {
        "assumption" | "hypothesis" => matches!(value.epistemic_state.as_str(), "open" | "testing"),
        "constraint" | "inference" => value.epistemic_state == "candidate",
        "decision" => value.epistemic_state == "proposed",
        "fact" => matches!(value.epistemic_state.as_str(), "candidate" | "contested"),
        "unknown" => matches!(value.epistemic_state.as_str(), "investigating" | "open"),
        _ => false,
    };
    if valid {
        Ok(())
    } else {
        Err(invalid("legacy epistemic_state is outside ADR-0047 states"))
    }
}

fn validate_body(
    value: &CognitiveAtom,
    allow_blank: bool,
) -> Result<(), KernelDecisionContractError> {
    if value.api_version != ATOM_API
        || value.canonicalization != CANONICALIZATION
        || value.effective_hardness != "none"
        || value.instruction_allowed
        || value.kind != ATOM_KIND
    {
        return Err(invalid("CognitiveAtom constants differ"));
    }
    identity(
        &value.atom_id,
        &value.atom_sha256,
        ATOM_PREFIX,
        "atom",
        allow_blank,
    )?;
    if !atom_type_member(&value.atom_type) {
        return Err(invalid("atom_type is unsupported"));
    }
    attestations(&value.attestations)?;
    bindings(&value.bindings)?;
    task(&value.task_binding)?;
    proposition(&value.proposition)?;
    scope(value)?;
    authority(&value.declared_authority)?;
    source(&value.source, &value.atom_type)?;
    validate_state(value)
}

fn validate_state(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    if value.validity.valid_from_unix_ms < 0
        || value
            .validity
            .valid_until_unix_ms
            .is_some_and(|end| end < 0 || end <= value.validity.valid_from_unix_ms)
    {
        return Err(invalid("CognitiveAtom validity is invalid"));
    }
    confidence(value)?;
    hardness(value)?;
    epistemic(value)
}

fn digest(value: &CognitiveAtom) -> Result<String, KernelDecisionContractError> {
    let mut blank = value.clone();
    blank.atom_id.clear();
    blank.atom_sha256.clear();
    validate_body(&blank, true)?;
    let canonical = wire::canonical_with_max(&blank, MAX_ATOM_BYTES)?;
    let mut hasher = Sha256::new();
    hasher.update(ATOM_DOMAIN);
    hasher.update(canonical.as_bytes());
    Ok(crate::governance_contract::codec::lower_hex(
        &hasher.finalize(),
    ))
}

/// Validates one exact, sealed `CognitiveAtom` v2.
///
/// # Errors
///
/// Returns an error for any wire-independent shape, semantic, bound, or digest violation.
pub fn validate_cognitive_atom(value: &CognitiveAtom) -> Result<(), KernelDecisionContractError> {
    validate_body(value, false)?;
    if value.atom_sha256 != digest(value)? {
        return Err(invalid("atom_sha256 does not match canonical preimage"));
    }
    wire::canonical_with_max(value, MAX_ATOM_BYTES)?;
    Ok(())
}

/// Seals one exact blank-identity `CognitiveAtom` v2 copy.
///
/// # Errors
///
/// Returns an error for nonblank identity or any invalid member, bound, or canonical preimage.
pub fn seal_cognitive_atom(
    value: &CognitiveAtom,
) -> Result<CognitiveAtom, KernelDecisionContractError> {
    if !value.atom_id.is_empty() || !value.atom_sha256.is_empty() {
        return Err(invalid("sealing CognitiveAtom requires blank identity"));
    }
    let mut sealed = value.clone();
    let digest = digest(&sealed)?;
    sealed.atom_id = format!("{ATOM_PREFIX}{digest}");
    sealed.atom_sha256 = digest;
    validate_cognitive_atom(&sealed)?;
    Ok(sealed)
}

/// Decodes exact compact canonical `CognitiveAtom` v2 bytes.
///
/// # Errors
///
/// Returns an error for malformed, noncanonical, oversized, semantically invalid, or unsealed bytes.
pub fn decode_cognitive_atom(bytes: &[u8]) -> Result<CognitiveAtom, KernelDecisionContractError> {
    let value = wire::decode_typed(bytes, MAX_ATOM_BYTES)?;
    validate_cognitive_atom(&value)?;
    Ok(value)
}

pub(super) fn validate_atoms(values: &[CognitiveAtom]) -> Result<(), KernelDecisionContractError> {
    if values.is_empty() || values.len() > 256 {
        return Err(invalid("cognitive_atoms cardinality must be 1..=256"));
    }
    for value in values {
        validate_cognitive_atom(value)?;
    }
    if !values
        .windows(2)
        .all(|pair| pair[0].atom_id.as_bytes() < pair[1].atom_id.as_bytes())
    {
        return Err(invalid(
            "cognitive_atoms must be strictly atom-id sorted and unique",
        ));
    }
    Ok(())
}
